### DISCUSSION TITLE
Is there any documentation on the compiler API?

### DISCUSSION THREAD

**Discussion by Zzzen** [microsoft/pyright#3407] [General]
like https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API

**Comment by erictraut** [comment]
Are you talking about the compiler API for the TypeScript compiler? You would need to look in the TypeScript documentation.

There is no public interface for the internal modules in pyright, so you won't find documentation outside of the source code.

**Reply by Zzzen** [reply]
Well, I mean the API of pyright. It can be used to develop tools like https://github.com/JoshuaKGoldberg/TypeStat and https://github.com/typescript-eslint/typescript-eslint/tree/main/packages/eslint-plugin.

**Reply by erictraut** [reply]
Pyright doesn't have a documented API. Its internal modules are all considered "implementation details" at this time. Formalizing some interface is something we might consider in the future, but I'm reluctant to do so now because it would be a support burden and limit our ability to evolve quickly. I regularly make large structural changes within pyright's code base for performance and maintainability reasons.

If you want to build another tool on top of pyright's core code, let's talk about what you're trying to do, and I can advise a way to do it in a way that will minimize the likelihood of breaking in the future.
